src/Microsoft.VisualStudio.LanguageServices.Razor/Editor/DefaultTextBufferFactoryService.cs

```diff
@@ -0,0 +1,141 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.ComponentModel.Composition;
+using System.Diagnostics;
+using Microsoft.VisualStudio.Editor.Razor;
+using Microsoft.VisualStudio.Text;
+
+namespace Microsoft.VisualStudio.LanguageServices.Razor.Editor
+{
+    [System.Composition.Shared]
+    [Export(typeof(TextBufferFactoryService))]
+    internal class DefaultTextBufferFactoryService : TextBufferFactoryService
+    {
+        private static readonly object RazorTextBufferInitializationKey = new object();
+
+        private readonly VisualStudioDocumentTrackerFactory _documentTrackerFactory;
+        private readonly VisualStudioRazorParserFactory _parserFactory;
+        private readonly BraceSmartIndenterFactory _braceSmartIndenterFactory;
+
+        [ImportingConstructor]
+        public DefaultTextBufferFactoryService(
+            VisualStudioDocumentTrackerFactory documentTrackerFactory,
+            VisualStudioRazorParserFactory parserFactory,
+            BraceSmartIndenterFactory braceSmartIndenterFactory)
+        {
+            if (documentTrackerFactory == null)
+            {
+                throw new ArgumentNullException(nameof(documentTrackerFactory));
+            }
+
+            if (parserFactory == null)
+            {
+                throw new ArgumentNullException(nameof(parserFactory));
+            }
+
+            if (braceSmartIndenterFactory == null)
+            {
+                throw new ArgumentNullException(nameof(braceSmartIndenterFactory));
+            }
+
+            _documentTrackerFactory = documentTrackerFactory;
+            _parserFactory = parserFactory;
+            _braceSmartIndenterFactory = braceSmartIndenterFactory;
+        }
+
+        public override bool TryGetDocumentTracker(ITextBuffer textBuffer, out VisualStudioDocumentTracker documentTracker)
+        {
+            if (textBuffer == null)
+            {
+                throw new ArgumentNullException(nameof(textBuffer));
+            }
+
+            if (!textBuffer.IsRazorBuffer())
```

Comment:
Could normalize all of these methods into a single generic method but wanted to keep them separate for clarity.

Reply:
The way you have it is best